### PR TITLE
Add test scenario specification for Zone11

### PR DESCRIPTION
**Pull request by marc-vanderwal**
## Purpose

This PR adds a test scenario specification for Zone11 that matches the test case specification as worded in #1415.

Of note here are the scenarios that perform tests on three different root zones. They will require three changes to the root hints during unit testing.

## Context

See #1256 (issue about Zone11).

Depends on #1415 (Zone11 test case specification update).

## Changes

- Add test scenario specification for Zone11.

## How to test this PR

Review.

### CODE REVIEW

File: docs/public/specifications/test-zones/Zone-TP/zone11.md [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+
+### Default zone configuration
+
+Unless otherwise specified in the specific scenario specification, the test zone
+for the scenario will follow the default setup as stated below. The `child zone`
+is the zone to be tested for the scenario.
+
+* The child zone is `SCENARIO.zone11.xa`.
+  * There is a zone file for the child zone.
+  * It is served by three name servers named ns1, ns2 and ns3.child.zone11.xa.
```
Comment (matsduf) [inline docs/public/specifications/test-zones/Zone-TP/zone11.md]: For the default scenarios I include one or two NS to make configuration smaller. What do you think will be gained by three?

File: docs/public/specifications/test-zones/Zone-TP/zone11.md [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+
+### SAME-SPF-DIFFERENT-TXT
+
+A zone whose SPF record is identical despite inconsistent TXT RRset records at
+the apex.
+
+* Zone: same-spf-different-txt.zone11.xa
+  * ns1 serves a TXT RRset containing one SPF record and one non-SPF record.
+  * ns2 and ns3 serve a different TXT RRset containing the same SPF record as
+    ns1.
```
Comment (matsduf) [inline docs/public/specifications/test-zones/Zone-TP/zone11.md]: Just one TXT record on ns2 and ns3, but two on ns1?

File: docs/public/specifications/test-zones/Zone-TP/zone11.md [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+  * It is served by ns-non-null-spf.root-servers.zone11.xa.
+  * The TXT RRset at the apex of the zone contains one TXT record with an SPF
+    record that isn’t null (i.e. permits at least one IP address).
+
+### INVALID-SYNTAX-1
+
+A zone whose SPF record is syntactically invalid.
+
+* Zone: invalid-syntax-1.zone11.xa
+  * The TXT RRset at the apex contains one SPF record with invalid syntax.
```
Comment (matsduf) [inline docs/public/specifications/test-zones/Zone-TP/zone11.md]: Does that mean that is starts with `v=spf1` but then contains something not permitted?

File: docs/public/specifications/test-zones/Zone-TP/zone11.md [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+
+### INVALID-SYNTAX-3
+
+A zone whose SPF record contains a few random bytes, with a few non-ASCII
+bytes among those.
+
+* Zone: invalid-syntax-2.zone11.xa
+  * The TXT RRset at the apex contains one SPF record with invalid syntax and
+    some random bytes outside the printable ASCII range.
```
Comment (matsduf) [inline docs/public/specifications/test-zones/Zone-TP/zone11.md]: Does that mean that is starts with v=spf1 but then contains something not permitted?

**Review by matsduf — DISMISSED** [review]
The test zones are not included.

**Comment by marc-vanderwal** [comment]
> The test zones are not included.

Yes, they’re coming, but I wanted some early feedback on the outline of the test scenarios before completing them.

File: docs/public/specifications/test-zones/Zone-TP/zone11.md (hunk 1 above)
Comment (marc-vanderwal) [reply docs/public/specifications/test-zones/Zone-TP/zone11.md]: There are two scenarios where I really need three: SPF-MISSING-ON-ONE and ALL-DIFFERENT-SPF. For all other scenarios, two are enough.

File: docs/public/specifications/test-zones/Zone-TP/zone11.md (hunk 2 above)
Comment (marc-vanderwal) [reply docs/public/specifications/test-zones/Zone-TP/zone11.md]: Yes, that’s right. I’ll clarify.

File: docs/public/specifications/test-zones/Zone-TP/zone11.md (hunk 3 above)
Comment (marc-vanderwal) [reply docs/public/specifications/test-zones/Zone-TP/zone11.md]: Yes, that’s right. I’ll clarify.

File: docs/public/specifications/test-zones/Zone-TP/zone11.md (hunk 4 above)
Comment (marc-vanderwal) [reply docs/public/specifications/test-zones/Zone-TP/zone11.md]: Yes, that’s right. I’ll clarify.

File: test-zone-data/Zone-TP/zone11/zone11.xa.zone [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+                        NS    ns2.invalid-syntax-2
+
+;; NOTE: ns2.invalid-syntax-2 deliberately has the same IPv4 and IPv6
+;; addresses as ns1.invalid-syntax-2. This is not a typo.
+ns1a.invalid-syntax-2   A     127.19.11.31
+                        AAAA  fda1:b2:c3::127:19:11:31
+ns1b.invalid-syntax-2   A     127.19.11.31
+                        AAAA  fda1:b2:c3::127:19:11:31
+ns2.invalid-syntax-2    A     127.19.11.32
+                        AAAA  fda1:b2:c3::127:19:11:32
```
Comment (matsduf) [inline test-zone-data/Zone-TP/zone11/zone11.xa.zone]: It looks like the text should read "**ns1a**.invalid-syntax-2 deliberately has the same IPv4 and IPv6 addresses as **ns1b**.invalid-syntax-2"

File: docs/public/specifications/test-zones/Zone-TP/zone11.md [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+  * The TXT RRset at the apex contains one SPF record with invalid syntax,
+    i.e. beginning with `v=spf1` followed by whitespace but not compliant with
+    the ABNF grammar described in [RFC 7208, Section 4.5][RFC 7208#4.5] and
+    following sections.
+
+### INVALID-SYNTAX-2
+
+Like INVALID-SYNTAX-1, but two of the name server names in the delegation and in
+the zone point to the same IP address. This scenario tests an edge case in the
+reporting of the name servers that serve an ill-formed SPF policy.
```
Comment (matsduf) [inline docs/public/specifications/test-zones/Zone-TP/zone11.md]: Is the purpose to verify that "ns-list" contains both names in spite of them having the same IP addresses?

File: test-zone-data/COMMON/arpa [hunk 7]
```diff
@@ -45,6 +45,10 @@ $ORIGIN 22.9.19.127.in-addr.arpa.
 $ORIGIN 2.2.0.0.9.0.0.0.9.1.0.0.7.2.1.0.0.0.0.0.3.c.0.0.2.b.0.0.1.a.d.f.ip6.arpa.
 @                       PTR  ns2.zone09.xa.
 
+$ORIGIN zone11.arpa.
+@                       NS   ns1.zone11.xa.
+@                       NS   ns2.zone11.xa.
+
```
Comment (matsduf) [inline test-zone-data/COMMON/arpa:51]: This is one way to do it. I have moved to creating test case dedicated root, xa and other zones instead to have as few shared files as possible to cut down the risk of conflicts

File: test-zone-data/Zone-TP/zone11/COMMON-SCENARIO-INCLUDE [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+                                 1h              ; retry
+                                 1w              ; expire
+                                 1d )            ; minimum
+
+                          NS     ns1.child.zone11.xa.
+                          NS     ns2.child.zone11.xa.
+
+;; Local Variables:
+;; mode: dns
+;; End:
```
Comment (matsduf) [inline test-zone-data/Zone-TP/zone11/COMMON-SCENARIO-INCLUDE:23]: Is this for Emacs?

**Review by matsduf — COMMENTED** [review]
Looks fine. The final test is when used as unit tests.

File: test-zone-data/Zone-TP/zone11/COMMON-SCENARIO-INCLUDE (hunk 8 above)
Comment (marc-vanderwal) [reply test-zone-data/Zone-TP/zone11/COMMON-SCENARIO-INCLUDE:23]: Yes, this is to switch to `dns-mode` instead of `fundamental-mode` when visiting the file in Emacs.

File: test-zone-data/Zone-TP/zone11/test-zones-output.md
Comment (tgreenx) [inline test-zone-data/Zone-TP/zone11/test-zones-output.md:1]: (File to be updated before merging)

File: docs/public/specifications/test-zones/Zone-TP/zone11.md [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+### NO-TXT
+
+A zone containing no TXT record at the apex.
+
+* Zone: no-txt.zone11.xa
+  * The TXT RRset at the apex is empty.
+
+### NO-SPF-TXT
+
+A zone containing at least one TXT record, but none of the TXT  after
```
Comment (tgreenx) [inline docs/public/specifications/test-zones/Zone-TP/zone11.md]:
```suggestion
A zone containing at least one TXT record, but none of the TXT after
```

File: docs/public/specifications/test-zones/Zone-TP/zone11.md [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+* Zone: no-spf-txt.zone11.xa
+  * The TXT RRset at the apex is non-empty, but none of the TXT records are
+    recognized as SPF records.
+
+### NO-SPF-ROOT-ZONE
+
+A root zone without SPF records.
+
+* Zone: "." (root zone)
+  * It is served by ns-no-spf.root-servers.zone11.xa.
```
Comment (tgreenx) [inline docs/public/specifications/test-zones/Zone-TP/zone11.md]:
```suggestion
  * It is served by ns1 and ns2.no-spf.root-servers.zone11.xa.
```

File: docs/public/specifications/test-zones/Zone-TP/zone11.md [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+* Zone: no-spf-arpa-zone.zone11.arpa
+  * It is served by ns1 and ns2.zone11.xa.
+  * The TXT RRset at the apex of the zone is empty.
+
+### NULL-SPF-ROOT-ZONE
+
+A root zone with a null SPF record.
+
+* Zone: "." (root zone)
+  * It is served by ns-null-spf.root-servers.zone11.xa.
```
Comment (tgreenx) [inline docs/public/specifications/test-zones/Zone-TP/zone11.md]:
```suggestion
  * It is served by ns1 and ns2.null-spf.root-servers.zone11.xa.
```

File: docs/public/specifications/test-zones/Zone-TP/zone11.md [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+  * It is served by ns1 and ns2.zone11.xa.
+  * The TXT RRset at the apex of the zone contains one TXT record with the
+    string `v=spf1 -all`.
+
+### NON-NULL-SPF-ROOT-ZONE
+
+A root zone with a non-null SPF record.
+
+* Zone: "." (root zone)
+  * It is served by ns-non-null-spf.root-servers.zone11.xa.
```
Comment (tgreenx) [inline docs/public/specifications/test-zones/Zone-TP/zone11.md]:
```suggestion
  * It is served by ns1 and ns2.non-null-spf.root-servers.zone11.xa.
```

File: docs/public/specifications/test-zones/Zone-TP/zone11.md [hunk 13]
```diff
@@ -0,0 +1,340 @@
...
+* Zone: no-spf-zone11
+  * It is served by ns1 and ns2.zone11.xa.
+  * The TXT RRset at the apex of the TLD zone is empty.
+
+### NO-SPF-ARPA-ZONE
+
+A zone in .arpa without SPF records.
+
+* Zone: no-spf-arpa-zone.zone11.arpa
+  * It is served by ns1 and ns2.zone11.xa.
```
Comment (tgreenx) [inline docs/public/specifications/test-zones/Zone-TP/zone11.md]:
```suggestion
  * It is served by ns1 and ns2.child.zone11.xa
```

File: docs/public/specifications/test-zones/Zone-TP/zone11.md [hunk 14]
```diff
@@ -0,0 +1,340 @@
...
+  * The TXT RRset at the apex of the TLD zone contains one TXT record with the
+    string `v=spf1 -all`.
+
+### NULL-SPF-ARPA-ZONE
+
+A zone in .arpa with a null SPF record.
+
+* Zone: null-spf-arpa-zone.zone11.arpa
+  * It is served by ns1 and ns2.zone11.xa.
```
Comment (tgreenx) [inline docs/public/specifications/test-zones/Zone-TP/zone11.md]:
```suggestion
  * It is served by ns1 and ns2.child.zone11.xa.
```

File: docs/public/specifications/test-zones/Zone-TP/zone11.md [hunk 15]
```diff
@@ -0,0 +1,340 @@
...
+  * The TXT RRset at the apex of the TLD zone contains one TXT record with an
+    SPF record that isn’t null (i.e. permits at least one IP address).
+
+### NON-NULL-SPF-ARPA-ZONE
+
+A zone in .arpa with a non-null SPF record.
+
+* Zone: non-null-spf-arpa-zone.zone11.arpa
+  * It is served by ns-non-null-spf.root-servers.zone11.xa.
```
Comment (tgreenx) [inline docs/public/specifications/test-zones/Zone-TP/zone11.md]:
```suggestion
  * It is served by ns1 and ns2.child.zone11.xa.
```

File: docs/public/specifications/test-zones/Zone-TP/zone11.md [hunk 16]
```diff
@@ -0,0 +1,340 @@
...
+  * The TXT RRset at the apex contains one SPF record with invalid syntax,
+    i.e. beginning with `v=spf1` followed by whitespace but not compliant with
+    the ABNF grammar described in [RFC 7208, Section 4.5][RFC 7208#4.5] and
+    following sections.
+
+### INVALID-SYNTAX-2
+
+Like INVALID-SYNTAX-1, but two of the name server names in the delegation and in
+the zone point to the same IP address. This scenario is there to ensure that
+the all name servers that serve an ill-formed SPF policy are reported, even if
```
Comment (tgreenx) [inline docs/public/specifications/test-zones/Zone-TP/zone11.md]:
```suggestion
all name servers that serve an ill-formed SPF policy are reported, even if
```

File: docs/public/specifications/test-zones/Zone-TP/zone11.md [hunk 17]
```diff
@@ -0,0 +1,340 @@
...
+    printable ASCII range, which is not compliant with the ABNF grammar
+    described in [RFC 7208, Section 4.5][RFC 7208#4.5] and following sections.
+
+### NON-AUTH-TXT
+
+A zone containing one SPF record in its TXT RRset, but served from name
+servers that fail to set the AA bit in their responses.
+
+* Zone: non-auth-txt.zone11.xa
+  * The TXT RRset at the apex contains one SPF record
```
Comment (tgreenx) [inline docs/public/specifications/test-zones/Zone-TP/zone11.md]:
```suggestion
  * The TXT RRset at the apex contains one SPF record.
```

File: docs/public/specifications/test-zones/Zone-TP/zone11.md [hunk 18]
```diff
@@ -0,0 +1,340 @@
...
+  * Both name servers serve the aforementioned TXT RRset with the AA bit unset
+    in the response header.
+
+### SERVFAIL
+
+A zone where queries for the TXT RRset are responded to with an RCODE
```
Comment (tgreenx) [inline docs/public/specifications/test-zones/Zone-TP/zone11.md]:
```suggestion
A zone where queries for the TXT RRset are responded with an RCODE
```

File: docs/public/specifications/test-zones/Zone-TP/zone11.md [hunk 19]
```diff
@@ -0,0 +1,340 @@
...
+different from NoError.
+
+* Zone: servfail.zone11.xa
+  * Queries for the TXT RRset are responded to with the [RCODE Name]
```
Comment (tgreenx) [inline docs/public/specifications/test-zones/Zone-TP/zone11.md]:
```suggestion
  * Queries for the TXT RRset are responded with the [RCODE Name]
```

File: docs/public/specifications/test-zones/Zone-TP/zone11.md [hunk 20]
```diff
@@ -0,0 +1,340 @@
...
+    "ServFail".
+
+### INCONSISTENT-SPF
+
+A zone that serves different SPF records depending on the name server being
+queried.
+
+* Zone: inconsistent-spf.zone11.xa
+  * ns1 serves a TXT RRset at the apex containing one SPF record
```
Comment (tgreenx) [inline docs/public/specifications/test-zones/Zone-TP/zone11.md]:
```suggestion
  * ns1 serves a TXT RRset at the apex containing one SPF record.
```

File: docs/public/specifications/test-zones/Zone-TP/zone11.md [hunk 21]
```diff
@@ -0,0 +1,340 @@
...
+  * ns2 serves a different TXT RRset at the apex containing one SPF record
+    that is also different from the SPF record served by ns1
```
Comment (tgreenx) [inline docs/public/specifications/test-zones/Zone-TP/zone11.md]:
```suggestion
    that is also different from the SPF record served by ns1.
```

File: docs/public/specifications/test-zones/Zone-TP/zone11.md [hunk 22]
```diff
@@ -0,0 +1,340 @@
...
+### SPF-MISSING-ON-ONE
+
+A zone that serves identical SPF records, except on one of the name servers.
+
+* Zone: spf-missing-on-one.zone11.xa
+  * The zone is delegated to three name servers named ns1, ns2 and
+    ns3.child.zone11.xa.
+  * ns2 and ns3 serve an identical TXT RRset at the apex containing one SPF
+    record
+  * ns1 serves a different TXT RRset at the apex without an SPF record
```
Comment (tgreenx) [inline docs/public/specifications/test-zones/Zone-TP/zone11.md]:
```suggestion
  * ns2 and ns3 serve an identical TXT RRset at the apex containing one SPF
    record.
  * ns1 serves a different TXT RRset at the apex without an SPF record.
```

File: docs/public/specifications/test-zones/Zone-TP/zone11.md [hunk 23]
```diff
@@ -0,0 +1,340 @@
...
+A zone that serves different SPF records on all of its name servers.
+
+* Zone: all-different-spf.zone11.xa
+  * The zone is delegated to three name servers named ns1, ns2 and
+    ns3.child.zone11.xa.
+  * ns1 serves a TXT RRset at the apex containing one SPF record
+  * ns2 serves a different TXT RRset at the apex containing one SPF record
+    that is also different
+  * ns3 serves another different TXT RRset at the apex containing one SPF
+    record that is different from the two previous records
```
Comment (tgreenx) [inline docs/public/specifications/test-zones/Zone-TP/zone11.md]:
```suggestion
  * ns1 serves a TXT RRset at the apex containing one SPF record.
  * ns2 serves a different TXT RRset at the apex containing one SPF record
    that is also different.
  * ns3 serves another different TXT RRset at the apex containing one SPF
    record that is different from the two previous records.
```

File: docs/public/specifications/test-zones/Zone-TP/zone11.md [hunk 24]
```diff
@@ -0,0 +1,340 @@
...
+
+### MULTIPLE-SPF-RECORDS
+
+A zone whose TXT RRset at the apex contains more than one SPF record.
+
+* Zone: multiple-spf-records.zone11.xa
+  * The TXT RRset at the apex contains more than one SPF record
```
Comment (tgreenx) [inline docs/public/specifications/test-zones/Zone-TP/zone11.md]:
```suggestion
  * The TXT RRset at the apex contains more than one SPF record.
```

File: docs/public/specifications/test-zones/Zone-TP/zone11.md [hunk 25]
```diff
@@ -0,0 +1,340 @@
...
+
+
+[Zone11]:                                 ../../tests/Zone-TP/zone11.md
+[RCODE Name]:                             https://www.iana.org/assignments/dns-parameters/dns-parameters.xhtml#dns-parameters-6
+[RFC 7208]:                               https://www.rfc-editor.org/rfc/rfc7208
+[RFC 7208#4.5]:                         https://www.rfc-editor.org/rfc/rfc7208#section-4.5
```
Comment (tgreenx) [inline docs/public/specifications/test-zones/Zone-TP/zone11.md]:
```suggestion
[RFC 7208#4.5]:                           https://www.rfc-editor.org/rfc/rfc7208#section-4.5
```

File: test-zone-data/Zone-TP/zone11/zone11.xa.zone [hunk 26]
```diff
@@ -0,0 +1,80 @@
...
+                        AAAA  fda1:b2:c3:0:127:19:11:22
+
+ns1.child               A     127.19.11.31
+                        AAAA  fda1:b2:c3::127:19:11:31
+ns2.child               A     127.19.11.32
+                        AAAA  fda1:b2:c3::127:19:11:32
+ns3.child               A     127.19.11.33
+                        AAAA  fda1:b2:c3::127:19:11:33
+
+;; TODO: do the same for all other scenarios in zone11.cfg
```
Comment (tgreenx) [inline test-zone-data/Zone-TP/zone11/zone11.xa.zone]: Can be removed?
```suggestion
```

**Review by tgreenx — COMMENTED** [review]
A few minor comments otherwise LGTM

File: docs/public/specifications/test-zones/Zone-TP/zone11.md (hunk 18 above)
Comment (marc-vanderwal) [reply docs/public/specifications/test-zones/Zone-TP/zone11.md]: It’s actually grammatically correct, but I’ll rephrase anyway.

File: test-zone-data/Zone-TP/zone11/zone11.xa.zone (hunk 26 above)
Comment (matsduf) [inline test-zone-data/Zone-TP/zone11/zone11.xa.zone]: Or specify what to do.

Comment (marc-vanderwal) [reply test-zone-data/Zone-TP/zone11/zone11.xa.zone]: It’s a remnant from a previous iteration where I didn’t have everything in the zone file. Now it is complete. I’ll remove it.

**Review by tgreenx — APPROVED** [review]

**Review by matsduf — COMMENTED** [review]
Looks fine.

**Review by matsduf — APPROVED** [review]

**Comment by matsduf** [comment]
@marc-vanderwal, please merge.
